### ISSUE TITLE
Inline SVG feature

### ISSUE THREAD

**Issue by joeldrapper** [joeldrapper/phlex#309]

**Comment by Bumppoman** [comment]
What are you envisioning for this?  I was just about to ask what the plan was for adding built-in support for the missing SVG tags but maybe there's a different direction?

**Comment by joeldrapper** [comment]
There are a couple of things I’d like to do for SVGs. First, I’d love to add an `inline_svg` helper based on this gem. https://github.com/jamesmartin/inline_svg

```ruby
inline_svg(file, **options)
```

I’d also be open to an alternate Phlex builder object, e.g. `Phlex::SVG` which has SVG elements instead of HTML elements.

```ruby
class MyImage < Phlex::SVG
  def template
    # svg tags available here
  end
end
```

Back in a `Phlex::View`, we could also update the `svg` tag block to yield a `Phlex::SVG` context.

```ruby
class MyView < Phlex::View
  def template
    svg do |svg|
      # svg tags available here on the yielded object, e.g.
      svg.circle
    end
  end
end
```

**Comment by Bumppoman** [comment]
Maybe allowing the class to be tuned instead of changing `Phlex::View`?  `View` is a natural fit whereas `HTML` sounds like a specialized library.

```ruby
class Page < Phlex::View
  # No annotation, so the class is initialized with the regular HTML tags
  def template
    div { 'Regular HTML page here.' }
  end
end

class Image < Phlex::View[:svg]
  # Now the class was initialized with the SVG tags instead of the HTML set
  def template
    svg do
      circle
    end
  end
end
```

**Comment by davekaro** [comment]
I'm in favor of something like `Phlex::SVG` to build your own SVG. Or, perhaps I misunderstand how the `inline_svg` helper would work. Would it depend on the https://github.com/jamesmartin/inline_svg gem or be entirely separate? In projects I work on now I create ViewComponent objects for icons, which enables me to pass CSS classes (just like inline_svg, without depending on inline_svg). Honestly, looking at inline_svg I may have misunderstood that you don't *need* to use the asset pipeline.

Basically, my usage is to render icons in SVG, and pass Taildwind CSS classes so I can style the icon however I need. Does that mean I should favor the `inline_svg` approach or the `Phlex::SVG` builder? What I like about using ViewComponents is they are no different from any other component. There is no caching of files read from disk or extra parsing steps. New up the object and it outputs SVG tags.

**Comment by joeldrapper** [comment]
I don’t think we should depend on the `inline_svg` gem as it has dependencies we don't need, but I do like the idea of being able to embed an SVG _file_ inline in a template.

**Comment by Bumppoman** [comment]
It just occurred to me that this builder approach could work for Turbo Stream content too.

**Comment by dhnaranjo** [comment]
I found myself inlining an SVG repeatedly in the same doc for reasons and looked into if I could use like, a <template/> tag or something and found this: https://stackoverflow.com/a/59614656. You can use the SVG `<use href="#another_svg_by_id" />` tag to inline once and reuse throughout. I wonder if an SVG Phlex component could like... tell if it has been called already and render a reference instead of another inlined copy.

**Comment by joeldrapper** [comment]
Rather than starting with `Phlex::HTML` and redeclaring all the elements, I want to make a more abstract `Phlex::XMarkupLanguage` which is like Phlex::HTML but doesn't have any elements or HTML-specific logic. `Phlex::SVG` and `Phlex::HTML` can inherit from that and register their own elements.

**Comment by hangsu** [comment]
That makes sense to me

**Comment by davekaro** [comment]
`Phlex::SVG` is on main now - https://github.com/joeldrapper/phlex/pull/493. No `inline_svg` but not sure we need it.

**Comment by joeldrapper** [comment]
Let's close this and open a new issue for the `inline_svg` feature.
